db/db_iter.cc

```diff
@@ -674,6 +692,11 @@ void DBIter::PrevInternal() {
       }
       return;
     }
+
+    if (TooManyInternalKeysSkipped()) {
+      return;
+    }
+
```

Comment:
If I understand the code correctly, valid_'s value need not always be the same as iter_->Valid()  value. In this code snippet, if valid_ is set to false in TooManyInternalKeysSkipped(), we will *not* break in 701 as we never enter the if block (iter_->Valid() need not be false). Hence a return here. 
In fact, some of my unit tests check this scenario.